fdbserver/BlobManager.actor.cpp

```diff
@@ -5868,10 +5868,14 @@ ACTOR Future<Void> watchBackupEnabled(Reference<BlobManagerData> bmData) {
 				if (lastFlushTs == 0) {
 					TraceEvent("StartInitialFlush").log();
 					loop {
-						Version lastFlushVersion = wait(maybeFlushGranules(bmData));
-						if (lastFlushVersion != invalidVersion) {
-							TraceEvent("CompleteInitialFlush").detail("Version", lastFlushVersion);
-							break;
+						try {
+							Version lastFlushVersion = wait(maybeFlushGranules(bmData));
+							if (lastFlushVersion != invalidVersion) {
+								TraceEvent("CompleteInitialFlush").detail("Version", lastFlushVersion);
+								break;
+							}
+						} catch (Error& e) {
+							TraceEvent(SevWarn, "InitialFlushError").error(e);
```

Comment:
we can't swallow all errors, this can cause serious problems, particularly if we swallow actor cancellations, internal errors, etc...
to be safe we only should swallow explicit known error types that are known to be non-fatal by this code.

Also don't we need to handle errors and retry in the non-initial flush as well?

Another improvement we talked about that we should implement instead of this is handle the retrying per-blob-range, and also on retry examine if the range is still blobbified/force purged before doing a retry, and abandon it if so

Reply:
> we can't swallow all errors
I can check error_code_operation_cancelled and throw it. Or I can only silent blob_granule_transaction_too_old,  blob_granule_request_failed etc that might be thrown from flush.. which one do you think better?

> don't we need to handle errors and retry in the non-initial flush as well?
for non-initial flush, I already log errors except for error_code_operation_cancelled.. see truncateMutationsLoop()

> ...examine if the range is still blobbified/force purged
From what I can see in the test env, it's a transient state. Even if I don't do any check but  just retry flush after a while, it succeeds.. I'm wonder if we really need to add more sophisticated logic here.

Reply:
We should already be retrying on blob_granule_request_failed internally, i think. If not then we should be.
And yeah the one error we specifically want to catch in this case is `blob_granule_transaction_too_old`.

The test environment is not conducive of the eventual desired environment though.
If we have many thousands of tenants on a cluster and each one is frequently dropping/creating databases, you will perpetually be in a state where this flush can never succeed.
And it's also just an efficiency thing too - if we flushed 999/1000 tenants and one hit this error, we would re-flush the other 999 tenants, creating extra work for hybrid and extra small delta files if we re-flushed every granule.
So it might be fine for now to only retry globally, but it's clearly desirable to handle this per-blobbified keyrange eventually as we scale the number of tenants and blob ranges in our clusters.

The ideal solution to solve both of those problems would be to have each range in maybeFlushBlobRange handle and retry this error separately, because the problem will always be isolated to a single blob range when it happens.

Reply:
just for record - discussed offline with Josh. it's better to add retry logic at range level..